<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείίγματος θα εξετάσετε μίία υλοποίηση με q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





<p>Πριν συνεχίσετε στην άσκηση να απαντήσετε στο εξής ερώτημα:</p>

<b><p>1. Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;</p></b>

* Ο αλγόριθμος Q learning βασίζεται στις τιμές των Q functions. Η τιμή αυτή δίνει για κάθε κατάσταση το αναμενόμενο κέρδος αν ακολουθήσουμε μια συγκεκριμένη κίνηση. 0 αλγόριθμος αυτός αποθηκεύει σε έναν πολυδιάστατο πίνακα τις τιμές των Q functions για όλες τις υποψήφιες καταστάσεις. Σε κάθε επανάληψη ακολουθεί την κίνηση που έχει το μέγιστο Q function και έχουμε ανανέωση της τιμής του Q function. Ο αλγόριθμος σε κάθε αρχικοποίηση του περιβάλλοντος έχει γνώση από το προηγούμενο επεισόδιο και με την ιδιότητα του Temporal Difference υπολογίζει εκ νέου τις Q function και η ανανέωση γίνεται με την διαφορά της νέας με την παλιά τιμή. Ύστερα απο πολλά επεισόδια πρέπει να παρατηρήσουμε σύγκλιση των τιμών.

* Η ενισχυτική μαθηση ταιριάζει σε προβήματα βελτιστοποίησης που προκύπτουν από μια ακολουθία ενεργειών

* Η βασική διαφορά του αλγορίθμου Q learning και Policy/Value Iteration είναι ότι δεν γνωρίζουμε εξ αρχής τις πιθανότητες μετάβασης και τις ανταμοιβές όπως επίσης o Q learning αλγόριθμος είναι  off policy σε αντίθεση με τους άλλους 2 που είναι on policy. Δηλαδή έχουμε τυχαία αρχικοποιήση θέσης και εξερευνούμε νέες πολιτικές ενώ σε αντίθεση με τους Policy/Value Iteration που έχουμε συνεχώς βελτίωση της αρχικής πολιτικής.

<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [1]:
!pip install cmake 'gym[atari]' scipy

In [3]:
import gym

env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [4]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [5]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [6]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

In [7]:
env.s = 328  # set environment to illustration's state

epochs = 0
penalties, reward = 0, 0

frames = [] # for animation

done = False

while not done:
    action = env.action_space.sample()
    state, reward, done, info = env.step(action)

    if reward == -10:
        penalties += 1
    
    # Put each rendered frame into dict for animation
    frames.append({
        'frame': env.render(mode='ansi'),
        'state': state,
        'action': action,
        'reward': reward
        }
    )

    epochs += 1
    
    
print("Timesteps taken: {}".format(epochs))
print("Penalties incurred: {}".format(penalties))

Timesteps taken: 3501
Penalties incurred: 1160


Παρατηρούμε ότι χρειάστηκαν πολλά timesteps και έχουμε πολλά penalties, συνεπώς δεν είναι ικανοποιητικά τα αποτελέσματα. Αυτό μπορεί να αποφευχθεί με την χρήση του reinforcement learning το οποίο μπορεί να παρέχει την γνώση από προηγούμενα επεισόδια.

<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>

<b><p>3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;</p></b>

* H παράμετρος α καθορίζει την συμβολή της προηγούμενης τιμής του Q function στην νέα τιμή. Αν πάρει την τιμή 1 τότε δεν υπολογίζεται καθόλου στην ανανέωση

* Η παράμετρος γ καθορίζει την συμβολή της μελλοντικής Q function, δηλαδή της επόμενης κίνησης. Όταν παίρνει την τιμή 1 τότε έχουμε την πλήρη συμβολή της.

In [8]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [9]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

CPU times: user 1min 10s, sys: 15.6 s, total: 1min 25s
Wall time: 1min 12s


In [10]:
q_table[328]

array([ -2.41113754,  -2.27325184,  -2.41179249,  -2.36119457,
       -11.03601551, -10.5252337 ])

In [19]:
"""Evaluate agent's performance after Q-learning"""

total_epochs, total_penalties,rewards_ep = 0,0,0,0
episodes = 100

for _ in range(episodes):
    state = env.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = env.step(action)

        if reward == -10:
            penalties += 1
        
        rewards_ep+=reward
        epochs += 1

    total_penalties += penalties
    total_epochs += epochs

print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average reward per move: {rewards_ep/total_epochs}")

Results after 100 episodes:
Average timesteps per episode: 13.07
Average penalties per episode: 0.0
Average reward per move: 0.6067329762815609


In [21]:
total_epochs, total_penalties,rewards_ep = 0,0,0

for _ in range(episodes):
  env.s = 328  # set environment to illustration's state

  epochs = 0
  penalties, reward = 0, 0

  frames = [] # for animation

  done = False

  while not done:
      action = env.action_space.sample()
      state, reward, done, info = env.step(action)

      if reward == -10:
          penalties += 1
      
      # Put each rendered frame into dict for animation
      frames.append({
          'frame': env.render(mode='ansi'),
          'state': state,
          'action': action,
          'reward': reward
          }
      )
      rewards_ep+=reward
      epochs += 1
    
  total_penalties += penalties
  total_epochs += epochs
      
      
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average reward per move: {rewards_ep/total_epochs}")

Results after 100 episodes:
Average timesteps per episode: 2003.02
Average penalties per episode: 644.53
Average reward per move: -3.8855278529420576


<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

Μέσος αριθμός παραβάσεων ανά επεισόδιο:

*   With Q learning : 0
*   Without Q learning : 644.53

Μέσος αριθμός βημάτων ανά διαδρομή:

*   With Q learning : 13.07
*   Without Q learning : 2003.02

Μέσος αριθμός ανταμοιβών ανά κίνηση:

*   With Q learning : 0.6067329762815609
*   Without Q learning : -3.8855278529420576



Παρατηρούμε πόσο υπερτερεί ο Q learning
